cnd/src/btsieve/bitcoin/bitcoin_cache.rs

```diff
@@ -0,0 +1,94 @@
+use crate::btsieve::{BlockByHash, LatestBlock};
+use bitcoin::{hashes::sha256d, util::hash::BitcoinHash};
+use fmt::Debug;
+use futures::Future;
+use futures_core::{
+    compat::Future01CompatExt,
+    future::{FutureExt, TryFutureExt},
+};
+use std::{fmt, fmt::Formatter, sync::Arc};
+use tokio::sync::Mutex;
+
+pub struct BitcoinCache<T> {
+    pub inner: T,
+    pub cache: Arc<Mutex<lru::LruCache<sha256d::Hash, bitcoin::Block>>>,
+}
+
+impl<T> Clone for BitcoinCache<T>
+where
+    T: Clone,
+{
+    fn clone(&self) -> Self {
+        BitcoinCache {
+            inner: self.inner.clone(),
+            cache: self.cache.clone(),
+        }
+    }
+}
+
+impl<T> Debug for BitcoinCache<T>
+where
+    T: Debug,
+{
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), fmt::Error> {
+        self.inner.fmt(f)
+    }
+}
+
+impl<T> LatestBlock for BitcoinCache<T>
+where
+    T: LatestBlock<Block = bitcoin::Block, BlockHash = sha256d::Hash> + Clone,
+{
+    type Block = bitcoin::Block;
+    type BlockHash = sha256d::Hash;
+
+    fn latest_block(
+        &mut self,
+    ) -> Box<dyn Future<Item = Self::Block, Error = anyhow::Error> + Send + 'static> {
+        let cache = self.cache.clone();
+        let mut inner = self.inner.clone();
+
+        let future = async move {
+            let block = inner.latest_block().compat().await?;
+            let mut guard = cache.lock().await;
+            guard.put(block.bitcoin_hash(), block.clone());
```

Comment:
Likely we should check for existence of the key before calling `put` here otherwise we clone the block and update the cache every call.  From docs for `LruCache`:
```
    /// Puts a key-value pair into cache. If the key already exists in the cache, then it updates
    /// the key's value and returns the old value. Otherwise, `None` is returned.
```